plugin/data_masking/src/udf/udf_gen_blacklist.cc

```diff
@@ -0,0 +1,120 @@
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software Foundation,
+   51 Franklin Street, Suite 500, Boston, MA 02110-1335 USA */
+
+#include "plugin/data_masking/include/udf/udf_gen_blacklist.h"
+#include "plugin/data_masking/include/plugin.h"
+#include "plugin/data_masking/include/udf/udf_utils.h"
+#include "plugin/data_masking/include/udf/udf_utils_string.h"
+
+#include <algorithm>
+
+static bool gen_blacklist_init(UDF_INIT *initid, UDF_ARGS *args,
+                               char *message) {
+  DBUG_ENTER("gen_blacklist_init");
+
+  if (args->arg_count != 3) {
+    std::snprintf(message, MYSQL_ERRMSG_SIZE,
+                  "Wrong argument list: gen_blacklist(str, dictionary_name, "
+                  "replacement_dictionary_name)");
+    DBUG_RETURN(true);
+  }
+
+  if (args->arg_type[0] != STRING_RESULT ||
+      args->arg_type[1] != STRING_RESULT ||
+      args->arg_type[2] != STRING_RESULT) {
+    std::snprintf(message, MYSQL_ERRMSG_SIZE,
+                  "Wrong argument type: gen_blacklist(string, string, string)");
+    DBUG_RETURN(true);
+  }
+
+  initid->maybe_null = 1;
+  initid->const_item =
+      0;  // Non-Deterministic: same arguments will produce different values
+  initid->ptr = NULL;
+
+  DBUG_RETURN(false);
+}
+
+static void gen_blacklist_deinit(UDF_INIT *initid) {
+  DBUG_ENTER("gen_blacklist_deinit");
+
+  if (initid->ptr) free(initid->ptr);
+
+  return;
+}
+
+/**
+ * Replaces a term present in one dictionary with a term from a second
+ * dictionary and returns the replacement term. This masks the original term by
+ * substitution.
+ *
+ * @param str: A string that indicates the term to replace.
+ * @param dictionary_name: A string that names the dictionary containing the
+ * term to replace.
+ * @param replacement_dictionary_name: A string that names the dictionary from
+ * which to choose the replacement term.
+ *
+ * @return A string randomly chosen from replacement_dictionary_name as a
+ * replacement for str, or str if it does not appear in dictionary_name, or NULL
+ * if either dictionary name is not in the dictionary registry. If the term to
+ * replace appears in both dictionaries, it is possible for the return
+ * value to be the same term.
+ */
+static std::string _gen_blacklist(const char *str, const char *dictionary_name,
+                                  const char *replacement_dictionary_name) {
+  std::string res(str);
+  std::string s_dictname_a(dictionary_name);
+  mysql::plugins::tolower(s_dictname_a);
+  std::string s_dictname_b(replacement_dictionary_name);
+  mysql::plugins::tolower(s_dictname_b);
+
+  // TODO: do we need to lock the mutex to read the data??? this would be a
```

Comment:
Yes, otherwise how do you prevent a concurrent removal of the dictionary?
Other simple idea to improve performance is to use rwlocks

Reply:
Thank you, done!